src/java.base/share/classes/jdk/internal/foreign/MemorySegmentRenderUtil.java

```diff
@@ -0,0 +1,367 @@
+                        state.appendIndex(index);
+                    }
+                    if (index % (HEX_STREAM_BYTES_PER_ROW >>> 1) == 0) {
+                        // We are either at the beginning or halfway through: add an extra space for readability
+                        state.appendSpace();
+                    }
+                    // Append the actual memory value
+                    state.appendValue(memorySegment.get(ValueLayout.JAVA_BYTE, index));
+                    final long nextCnt = index + 1;
+                    if (nextCnt % HEX_STREAM_BYTES_PER_ROW == 0 || nextCnt == memorySegment.byteSize()) {
+                        // We have a complete line (eiter a full line or the last line)
+                        return state.renderLineToStringAndReset();
+                    } else {
+                        // For this count, there was no line break so pass null and filter it away later
+                        return null;
+                    }
+                })
+                .filter(Objects::nonNull);
+    }
+
+    /**
+     * Returns a human-readable view of the provided {@code memorySegment} viewed through
+     * the provided {@code lens}.
+     * <p>
+     * Lines are separated with the system-dependent line separator {@link System#lineSeparator() }.
+     * Otherwise, the exact format of the returned view is unspecified and should not
+     * be acted upon programmatically.
+     * <p>
+     * As an example, a MemorySegment viewed though the following memory layout
+     * {@snippet lang = java:
+     * var lens = MemoryLayout.structLayout(
+     *         ValueLayout.JAVA_INT.withName("x"),
+     *         ValueLayout.JAVA_INT.withName("y")
+     * ).withName("Point");
+     *}
+     * might be rendered to something like this:
+     * {@snippet lang = text:
+     * Point {
+     *   x=1,
+     *   y=2
+     * }
+     *}
+     * <p>
+     * This method is intended to view memory segments through small and medium-sized memory layouts
+     * and is, in all cases, restricted by the inherent String capacity limit.
+     *
+     * @param memorySegment to be viewed
+     * @param lens  to use as a lens when viewing the memory segment
+     * @return a view of the memory segment viewed through the memory layout
+     * @throws OutOfMemoryError if the view exceeds the array size VM limit
+     */
+    public static String viewThrough(MemorySegment memorySegment,
+                                     MemoryLayout lens) {
+        requireNonNull(memorySegment);
+        requireNonNull(lens);
+
+        final var sb = new StringBuilder();
+        final Consumer<CharSequence> action = line -> {
+            if (!sb.isEmpty()) {
+                sb.append(System.lineSeparator());
+            }
+            sb.append(line);
+        };
+        renderView(memorySegment, lens, action, new ViewState(), "");
+        return sb.toString();
+    }
+
+    public static void renderView(MemorySegment memorySegment,
```

Comment:
Suggestion, instead of keeping track of an explicit index, it would be perhaps better to construct a layout path using `MemoryLayout.PathElement`. That way, access to the correct element can be made w/o index. It will be slower, as it will compute a new VH each time, but at least it's guaranteed to be correct. Another approach could be to construct the layout path, and then use it to compute the offset, and then use the offset as you are doing now. I'm worried about duplicating offsetting logic.